File: Services/pkg/domain/role_permission.go
package domain

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type RolePermissionRepository interface {
	Migrate() error
	CreateRolePermission(ctx context.Context, rolePermission models.RolePermission) *helpers.ResponseError
	GetRolePermission(ctx context.Context, id uint) (*models.RolePermission, *helpers.ResponseError)
	GetRolePermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.RolePermission, *models.Pagination, *models.Search, *helpers.ResponseError)
	GetRolePermissionsByRoleID(ctx context.Context, roleID uint) ([]models.RolePermission, *helpers.ResponseError)
	UpdateRolePermission(ctx context.Context, id uint, rolePermission models.RolePermission) *helpers.ResponseError
	DeleteRolePermission(ctx context.Context, id uint) *helpers.ResponseError
	DeleteRolePermissionsByRoleID(ctx context.Context, roleID uint) *helpers.ResponseError
}

type RolePermissionService interface {
	CreateRolePermission(ctx context.Context, rolePermission models.RolePermission) []helpers.ResponseError
	GetRolePermission(ctx context.Context, id uint) (*models.RolePermission, []helpers.ResponseError)
	GetRolePermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.RolePermission, *models.Pagination, *models.Search, []helpers.ResponseError)
	GetRolePermissionsByRoleID(ctx context.Context, roleID uint) ([]models.RolePermission, []helpers.ResponseError)
	UpdateRolePermission(ctx context.Context, id uint, rolePermission models.RolePermission) []helpers.ResponseError
	DeleteRolePermission(ctx context.Context, id uint) []helpers.ResponseError
	DeleteRolePermissionsByRoleID(ctx context.Context, roleID uint) []helpers.ResponseError
}

File: hono-clean-architecture/pkg/domain/role_permission.ts
import { RolePermission, Pagination, Search } from "../models";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export interface RolePermissionRepository {
  migrate(): Promise<void>;
  createRolePermission(rolePermission: {
    roleId: number;
    permissionId: number;
  }): Promise<ResponseError | null>;
  getRolePermission(
    id: number
  ): Promise<{ rolePermission: RolePermission | null; error: ResponseError | null }>;
  getRolePermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    rolePermissions: RolePermission[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }>;
  deleteRolePermission(id: number): Promise<ResponseError | null>;
  getRolePermissionsByRoleId(
    roleId: number
  ): Promise<{ rolePermissions: RolePermission[]; error: ResponseError | null }>;
}

export interface RolePermissionService {
  createRolePermission(rolePermission: {
    roleId: number;
    permissionId: number;
  }): Promise<ResponseError[]>;
  getRolePermission(
    id: number
  ): Promise<{ rolePermission: RolePermission | null; errors: ResponseError[] }>;
  getRolePermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    rolePermissions: RolePermission[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }>;
  deleteRolePermission(id: number): Promise<ResponseError[]>;
}
